### PR TITLE
Advanced Room Movement - Costmatrixes, Source Keeper Avoidance, Hostile Room Avoidance

### LINKED ISSUE #87: [feature] Advanced Multiroom Travel
- [ ] Avoid enemy rooms (unless desired).
- [ ] Generate custom costmatrixes with features such as spot avoidance (for walking around source keepers or towers).
- [x] Detect when stuck and compensate.
- [x] Switch `ignoreCreeps` on and off depending on circumstances.

### PR DESCRIPTION
**Pull request by tedivm**
This adds costmatrix support to the system. This gives `travelTo` some new options-

* `avoid` - don't walk on these positions

* `avoidRange` - Each item is an object containing a `pos`, `range` and optionally `value` (defaults to 255). Every position within the `range` of the supplied `pos` will be given `value` cost.

* `ignore` - only use terrain costs on these positions.

* `ignoreSourceKeepers` - by default this version moves creeps around where source keepers normally are, but this option will disable that behavior.

* `ignoreHostileCities` - by default creeps will avoid rooms that are owned by other players. Use this to disable that.

* `ignoreHostileReservations` - by default creeps will avoid rooms that are reserved by other players. Use this to disable that.

Some other behavioral notes-

* CostMatrixes are cached with a max age of 3000 ticks. If the room is visible though it will have a ttl of 25 ticks. This keeps costmatrixes from getting too stale while also also making it so lack of visibility in a room doesn't harm pathfinding.

* Room ownership and levels are pulled from the `Intel` system.

resolves #87

### CODE REVIEW

File: src/extends/room/movement.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  // Add in creeps
+  if (!opts.ignoreCreeps && Game.rooms[roomname]) {
+    const room = Game.rooms[roomname]
+    const creeps = room.find(FIND_CREEPS)
+    for (let creep of creeps) {
+      cm.set(creep.pos.x, creep.pos.y, 255)
+    }
+  }
+
+  // Set these positions to just use terrain score
```
Comment (Esryok) [inline src/extends/room/movement.js]: Currently 'avoidRange' & 'avoid' both trump 'ignore'. Deliberate?

File: src/extends/room/movement.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  if (!opts.ignoreCreeps && Game.rooms[roomname]) {
+    const room = Game.rooms[roomname]
+    const creeps = room.find(FIND_CREEPS)
+    for (let creep of creeps) {
+      cm.set(creep.pos.x, creep.pos.y, 255)
+    }
+  }
+
+  // Block all of these positions as unwalkable
+  if (opts.avoid) {
```
Comment (TheRealMaxion) [inline src/extends/room/movement.js]: Should perhaps this also be moved below the application of avoidRange, since avoidRange can set lower values?

Comment (tedivm) [reply src/extends/room/movement.js]: I did it about two seconds before you made this comment here :-)

Comment (TheRealMaxion) [reply src/extends/room/movement.js]: Good, just figured I'd put my comment into github as well to be thorough for posterity. :)

**Pull request merged**